http/src/main/resources/jupyter/gce-init.sh

```diff
@@ -0,0 +1,528 @@
+function retry {
+  local retries=$1
+  shift
+
+  for ((i = 1; i <= $retries; i++)); do
+    # run with an 'or' so set -e doesn't abort the bash script on errors
+    exit=0
+    "$@" || exit=$?
+    if [ $exit -eq 0 ]; then
+      return 0
+    fi
+    wait=$((2 ** $i))
+    if [ $i -eq $retries ]; then
+      log "Retry $i/$retries exited $exit, no more retries left."
+      break
+    fi
+    log "Retry $i/$retries exited $exit, retrying in $wait seconds..."
+    sleep $wait
+  done
+  return 1
+}
+
+function log() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@"
+}
+
+#####################################################################################################
+# Main starts here. It is composed of three sections:
+#   1. Set up that is NOT specific to GCE_OPERATION
+#   2. Operations we want to perform only when we are 'creating' a VM
+#   3. Operations we want to perform only when we are 'restarting' a VM that was previously created
+#####################################################################################################
+
+# Array for instrumentation
+# UPDATE THIS IF YOU ADD MORE STEPS:
+# currently the steps are:
+# START init,
+# .. after env setup
+# .. after copying files from google and into docker
+# .. after docker compose
+# .. after welder start
+# .. after hail and spark
+# .. after nbextension install
+# .. after server extension install
+# .. after combined extension install
+# .. after user script
+# .. after lab extension install
+# .. after jupyter notebook start
+# END
+
+
+#####################################################################################################
+# Set up that is NOT specific to GCE_OPERATION
+#####################################################################################################
+STEP_TIMINGS=($(date +%s))
+
+GCE_OPERATION=$(gceOperation)
+JUPYTER_HOME=/etc/jupyter
+
+log "Running GCE VM init script in ${GCE_OPERATION} mode..."
+
+export CLUSTER_NAME=$(clusterName)
+export GOOGLE_PROJECT=$(googleProject)
+export STAGING_BUCKET=$(stagingBucketName)
+export OWNER_EMAIL=$(loginHint)
+export JUPYTER_SERVER_NAME=$(jupyterServerName)
+export JUPYTER_DOCKER_IMAGE=$(jupyterDockerImage)
+export JUPYTER_START_USER_SCRIPT_URI=$(jupyterStartUserScriptUri)
+# Include a timestamp suffix to differentiate different startup logs across restarts.
+export JUPYTER_START_USER_SCRIPT_OUTPUT_URI="$(jupyterStartUserScriptOutputBaseUri)-$(date -u "+%Y.%m.%d-%H.%M.%S").txt"
+export NOTEBOOKS_DIR=$(notebooksDir)
+export WELDER_SERVER_NAME=$(welderServerName)
+export WELDER_DOCKER_IMAGE=$(welderDockerImage)
+export WELDER_ENABLED=$(welderEnabled)
+
+# If a Google credentials file was specified, grab the service account json file
+# and set the GOOGLE_APPLICATION_CREDENTIALS EV.
+# This overrides the credentials on the metadata server.
+# SERVICE_ACCOUNT_CREDENTIALS=$(jupyterServiceAccountCredentials)
+SERVICE_ACCOUNT_CREDENTIALS=""
+if [ ! -z ${SERVICE_ACCOUNT_CREDENTIALS} ] ; then
```

Comment:
nit: I was curious about the ${var} syntax, vs $var and "$var", so I did a bit of digging. Apparently, it's good practice to surround this in quotes, and I think the ${} is unecessary

i.e. `"$SERVICE_ACCOUNT_CREDENTIALS"`

https://stackoverflow.com/questions/18135451/what-is-the-difference-between-var-var-and-var-in-the-bash-shell

tl;dr ${} is only needed if its part of a larger string, "$var" is good to use when invoking control statements on the variable to avoid unintended effects

Reply:
Cool - thanks for the reference!

Reply:
Cool; I can never remember this stuff